File: lagomtest/household-impl/src/main/java/household/impl/HouseholdServiceImpl.java
package household.impl;

import java.util.UUID;

import javax.inject.Inject;

import com.lightbend.lagom.javadsl.api.ServiceCall;
import com.lightbend.lagom.javadsl.api.transport.NotFound;
import com.lightbend.lagom.javadsl.persistence.PersistentEntityRef;
import com.lightbend.lagom.javadsl.persistence.PersistentEntityRegistry;

import akka.Done;
import akka.NotUsed;
import household.api.Household;
import household.api.HouseholdService;
import household.impl.HouseholdCommand.GetHousehold;
import household.impl.HouseholdCommand.PutHousehold;

public class HouseholdServiceImpl implements HouseholdService {

	@Inject
	private PersistentEntityRegistry persistentEntityRegistry;

	@Inject
	public HouseholdServiceImpl(PersistentEntityRegistry persistentEntityRegistry) {
		this.persistentEntityRegistry = persistentEntityRegistry;
		persistentEntityRegistry.register(HouseholdEntity.class);
	}

	@Override
	public ServiceCall<NotUsed, Household> getHousehold(UUID householdId) {
		return request -> {
			// Look up the hello world entity for the given ID.
			PersistentEntityRef<HouseholdCommand> ref = persistentEntityRegistry.refFor(HouseholdEntity.class, householdId.toString());
			// Ask the entity the ReadHousehold command.
			return ref
					.ask(new GetHousehold())
					.thenApply(maybeHousehold -> {
						if (maybeHousehold.isPresent()) {
							return maybeHousehold.get();
						} else {
							throw new NotFound("Household " + householdId + " not found");
						}
					});
		};
	}

	@Override
	public ServiceCall<Household, Done> putHousehold() {
		return household -> {
			PersistentEntityRef<HouseholdCommand> ref = persistentEntityRegistry.refFor(HouseholdEntity.class, household.id().toString());
			return ref.ask(new PutHousehold(household));
		};
	}

}

File: lagomscala/household-impl/src/main/scala/household/impl/HouseholdServiceImpl.scala
package household.impl

import akka.actor.ActorSystem
import akka.stream.Materializer
import akka.{Done, NotUsed}
import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.scaladsl.persistence.PersistentEntityRegistry
import lagomscala.household.api.{Household, HouseholdService, InventoryItem}
import java.util.UUID

import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.scaladsl.api.transport.{NotFound, PolicyViolation, TransportErrorCode, TransportException}
import com.lightbend.lagom.scaladsl.persistence.PersistentEntity.ReplyType
import com.lightbend.lagom.scaladsl.persistence.PersistentEntityRegistry

import scala.collection.immutable
import scala.collection.immutable.Seq
import scala.concurrent.ExecutionContext

/**
  * Created by srn7919 on 12.08.17.
  */
class HouseholdServiceImpl(registry: PersistentEntityRegistry, system: ActorSystem)(implicit ec: ExecutionContext, mat: Materializer) extends HouseholdService {

  override def create = ServiceCall { household =>
    val id = UUID.randomUUID()
    val newHousehold = household.copy(id = Some(id))
    refFor(id).ask(CreateHousehold(newHousehold)).map { _ => newHousehold }
  }

  override def getOne(householdId: UUID): ServiceCall[NotUsed, Household] = ServiceCall { _ =>
    refFor(householdId).ask(GetHousehold).map {
      case Some(household) => household
      case None => throw NotFound(s"Household with id ${householdId.toString} does not exist!")
    }
  }

  override def consume(householdId: UUID, itemId: UUID) = ServiceCall[Float, Done] { newRemainingFraction =>
    refFor(householdId).ask(Consume(itemId, newRemainingFraction))
  }

  override def stock(householdId: UUID): ServiceCall[InventoryItem, Done] = ServiceCall { item =>
    refFor(householdId).ask(Stock(item))
  }

  override def inventory(householdId: UUID) = ServiceCall[NotUsed, Seq[InventoryItem]] { _ =>
    refFor(householdId).ask(GetInventoryItems)
  }

  private def refFor(householdId: UUID) = registry.refFor[HouseholdEntity](householdId.toString)
}
